code/test_dependency_injection.cpp

```diff
@@ -0,0 +1,84 @@
+#include <iostream>
+#include <cmath>
+#include "gtest/gtest.h"
+
+/// @brief A "complex" calculation of the norm.
+/// @param array The vector to calculate the norm for.
+/// @return The actual norm of the vector.
+double calculate_norm(std::vector<int> array)
+{
+  double output{0};
+  for (int i{0}; i < array.size(); ++i)
+  {
+    output = array[i] * array[i];
+  }
+  return std::sqrt(output);
+}
+
+/// @brief Just a stub, so we can get a simple value to test things with.
+/// @param array The vector to calculate the norm for.
+/// @return The norm, a dummy value in this case.
+double norm_stub(std::vector<int> array)
+{
+  return 10.0;
+}
+
+/// @brief We cannot do dependency injection, so the result depends on what calculate_norm is doing
+/// @param array The vector to normalize.
+void normalize_v1(std::vector<int> array)
+{
+  double norm{calculate_norm(array)};
+
+  for (int i{0}; i < array.size(); ++i)
+  {
+    array[i] /= norm;
+  }
+}
+
+/// @brief The norm calculation is injected, so we control that value.
+/// @param array The vector to normalize.
+/// @param func The function used to calculate the norm.
+void normalize_v2(
+    std::vector<int> array,
+    std::function<double(std::vector<int>)> func = calculate_norm)
+{
+  double norm{func(array)};
+
+  for (int i{0}; i < array.size(); ++i)
+  {
+    array[i] /= norm;
+  }
+}
+
+TEST(NormalizeTest, WithoutDependencyInjection)
+{
+  std::vector<int> input{1, 2, 3};
+
+  // We need the exact value of the norm in order to check that
+  // the calculation is correct.
+  double factor{calculate_norm(input)};
+  std::vector<double> copy{1 / factor, 2 / factor, 3 / factor};
```

Comment:
Instead of manually dividing each element, please make use of a function or loop as shown below.
```suggestion
  void divide_each_element(std::vector<int> &vec, int factor)
{
    for (auto &elem : vec)
        elem /= factor;
}
```

Reply:
For actual code, definetly I would go with this - it is what is done in the normalise function. But for the test case I think it is best to keep it simple and explicit, and also calculated in a diferent way than in the normalize function, so you can asses you're doing it right as are getting at the same result in different ways.

That's a general advise for tests: when possible (and is not always possible), calculate the expected result in a different way that what the code under test is doing, so you have an extra reasurance that the code is working.

Reply:
@dalonsoa I agree partially. I agree that it might be a good idea to check results in different ways. However, the part highlighted in the code is not fundamentally doing anything different than if you use a loop.

One is a good practice while other is not (even if number of elements are small).

I leave the final decision on you. Please feel free to keep it as it is.

Reply:
Fixed in f49a240